doc/src/manual/methods.md

```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
 "definition for Int"
 ```
 
+## Design Patterns with Parametric Methods
+
+
+While complex dispatch logic is not required for performance or usability,
+sometimes it can be the best way to express some algorithm.
+Here are a few common design patterns that come up sometimes when using dispatch in this way.
+
+### Extracting the type parameter from a super-type
+
+
+Here is the correct code template for returning the element-type `T`
+of any arbitrary subtype of `AbstractArray`:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{<:AbstractArray{T}}) where {T} = T
+```
+using so-called triangular dispatch.
+
+Another way, which used to be the only correct way before the advent of
+triangular dispatch, is:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{AbstractArray}) = Any
+eltype{T}(::Type{AbstractArray{T}}) = T
+eltype{T, N}(::Type{AbstractArray{T, N}}) = T
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
```

Comment:
> It doesn't return `Integer` because the `eltype` is a definite subtype of `Integer`, but it is not `Integer`.

There is also this case:
```
julia> eltype(Array{T, 1} where T<:Int)                                                                                                               
Any                                                                                                                                                   
```
where the eltype must be `Int`.  This could be fixed reasonably easy if `eltype(Array{T, 1} where T<:Integer)` would return `Integer`.